pkg/utils/addons/addons.go

```diff
@@ -0,0 +1,166 @@
+package addons
+
+import (
+	mfc "github.com/manifestival/controller-runtime-client"
+	mf "github.com/manifestival/manifestival"
+	"github.com/tektoncd/operator/pkg/flag"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type pipelineGenerator interface {
+	generate(pipeline unstructured.Unstructured) (unstructured.Unstructured, error)
+}
+
+type dcPipeline struct {
+	environment string
+	nameSuffix  string
+}
+
+type deploymentPipeline struct {
+	environment string
+	nameSuffix  string
+}
+
+type knativePipeline struct {
+	environment string
+	nameSuffix  string
+}
+
+func (dcp *dcPipeline) generate(pipeline unstructured.Unstructured) (unstructured.Unstructured, error) {
+	newTempRes := unstructured.Unstructured{}
+	pipeline.DeepCopyInto(&newTempRes)
+	labels := newTempRes.GetLabels()
+	labels[flag.LabelEnvironmentType] = dcp.environment
+	newTempRes.SetLabels(labels)
```

Comment:
this par is common in all generate functions..

we just need to move `the parts which` are different into individual methods

Reply:
So all the generating functions are adding third task which is **deploy** in its own way, we do see some of the similar parts in that task. And I wanted to keep this task separately. I will check if I can move things to a different function, something like which generate the base of the task.

Reply:
the three struct has the same fields
```
environment string
nameSuffix  string
```
then why do we need three structs for same type of objects...


so what we need, is one type (`struct`) with the same fields 
```
environment string
	nameSuffix  string

Reply:
atleast, let us create a type `pipeline`
```
type pipeline struct {
       environment string
	nameSuffix  string
}
```

then embedd it

eg: 
```
type deploymentPipeline struct {
	pipeline
}
```

Reply:
or what we could do is 

create a type
```
type Pipeline struct {
       environment string
	nameSuffix  string
       pipelineGenerator
}
```

```
eg: 
type dcGenerator struct {}

func (d dcGenerator) generate() {
}
```

the `pipelineGenerator` type could be `interface type` or a `function type`

and then in create pipelines

```pipelines := []pipelines{
		&Pipeline{environment: "openshift", nameSuffix: "", dcGenerator{}},
		&Pipeline{environment: "kubernetes", nameSuffix: "-deployment", deploymentGenerator},
		&Pipeline{environment: "knative", nameSuffix: "-knative", knativeGenerator},
	}
```